src/Decoder.c

```diff
@@ -4320,6 +4201,208 @@ static ZyanStatus ZydisNodeHandlerMvexE(const ZydisDecodedInstruction* instructi
 
 /* ---------------------------------------------------------------------------------------------- */
 
+/**
+ * Populates the internal register id fields for `REG`, `RM`, `NDSNDD`, `BASE` and `INDEX`/`VIDX`
+ * encoded operands and performs sanity checks.
+ *
+ * @param   context     A pointer to the `ZydisDecoderContext` instance.
+ * @param   instruction A pointer to the `ZydisDecodedInstruction` struct.
+ * @param   def_reg     The type definition for the `.reg` encoded operand.
+ * @param   def_rm      The type definition for the `.rm` encoded operand.
+ * @param   def_ndsndd  The type definition for the `.vvvv` encoded operand.
+ *
+ * @return  A zyan status code.
+ *
+ * This function sets all unused register ids to `-1`. This rule does currently not apply to
+ * `base` and `index`.
+ *
+ * Definition encoding:
+ * - `def_reg`    -> `ZydisRegisterKind`
+ * - `def_ndsndd` -> `ZydisRegisterKind`
+ * - `def_rm`     -> `ZydisRegisterKind` (`.mod == 3`) or ZydisMemoryOperandType (`.mod != 3`)
+ */
+static ZyanStatus ZydisPopulateRegisterIds(ZydisDecoderContext* context,
+    const ZydisDecodedInstruction* instruction, ZyanU8 def_reg, ZyanU8 def_rm, ZyanU8 def_ndsndd)
+{
+    ZYAN_ASSERT(context);
+    ZYAN_ASSERT(instruction);
+
+    const ZyanBool is_reg = context->reg_ids.is_reg;
+    const ZyanBool has_sib = !is_reg && (instruction->raw.modrm.rm != 4);
+    const ZyanBool has_vsib = has_sib && (def_rm == ZYDIS_MEMOP_TYPE_VSIB);
+
+    ZyanU8 id_reg    = instruction->raw.modrm.reg;
+    ZyanU8 id_rm     = instruction->raw.modrm.rm;
+    ZyanU8 id_ndsndd = context->cache.vvvv;
```

Comment:
This doesn't work as intended in 16/32-bit mode because highest bit of `id_ndsndd` might be already set and nothing is done to reset it in 16/32-bit mode (but there's code to set it in 64-bit mode).
```suggestion
    ZyanU8 id_ndsndd = context->cache.vvvv & 7;
```

Reply:
If we do it like this, we additionally have to restore bit 4 in the 64 bit mode branch, I guess.

Reply:
Ah, true. I forgot 64-bit branch sets even higher bit only.